source/common/config/delta_subscription_state.h

```diff
@@ -1,55 +1,55 @@
 #include "common/common/logger.h"
-#include "common/config/pausable_ack_queue.h"
+#include "common/config/subscription_state.h"
+
+#include "absl/types/optional.h"
 
 namespace Envoy {
 namespace Config {
 
-// Tracks the xDS protocol state of an individual ongoing delta xDS session, i.e. a single type_url.
-// There can be multiple DeltaSubscriptionStates active. They will always all be
-// blissfully unaware of each other's existence, even when their messages are
-// being multiplexed together by ADS.
-class DeltaSubscriptionState : public Logger::Loggable<Logger::Id::config> {
+// Tracks the state of a delta xDS-over-gRPC protocol session.
+class DeltaSubscriptionState : public SubscriptionState {
 public:
   DeltaSubscriptionState(std::string type_url, SubscriptionCallbacks& callbacks,
-                         const LocalInfo::LocalInfo& local_info,
                          std::chrono::milliseconds init_fetch_timeout,
                          Event::Dispatcher& dispatcher);
+  ~DeltaSubscriptionState() override;
 
   // Update which resources we're interested in subscribing to.
   void updateSubscriptionInterest(const std::set<std::string>& cur_added,
-                                  const std::set<std::string>& cur_removed);
+                                  const std::set<std::string>& cur_removed) override;
 
   // Whether there was a change in our subscription interest we have yet to inform the server of.
-  bool subscriptionUpdatePending() const;
+  bool subscriptionUpdatePending() const override;
 
-  void markStreamFresh() { any_request_sent_yet_in_current_stream_ = false; }
+  void markStreamFresh() override { any_request_sent_yet_in_current_stream_ = false; }
 
-  UpdateAck handleResponse(const envoy::api::v2::DeltaDiscoveryResponse& message);
+  // message is expected to be a envoy::api::v2::DeltaDiscoveryResponse.
+  UpdateAck handleResponse(const void* response_proto_ptr) override;
 
-  void handleEstablishmentFailure();
+  void handleEstablishmentFailure() override;
 
   // Returns the next gRPC request proto to be sent off to the server, based on this object's
   // understanding of the current protocol state, and new resources that Envoy wants to request.
-  envoy::api::v2::DeltaDiscoveryRequest getNextRequestAckless();
+  // Returns a new'd pointer, meant to be owned by the caller.
+  void* getNextRequestAckless() override;
   // The WithAck version first calls the Ackless version, then adds in the passed-in ack.
-  envoy::api::v2::DeltaDiscoveryRequest getNextRequestWithAck(const UpdateAck& ack);
+  // Returns a new'd pointer, meant to be owned by the caller.
+  void* getNextRequestWithAck(const UpdateAck& ack) override;
 
   DeltaSubscriptionState(const DeltaSubscriptionState&) = delete;
   DeltaSubscriptionState& operator=(const DeltaSubscriptionState&) = delete;
 
 private:
+  // Returns a new'd pointer, meant to be owned by the caller.
+  envoy::api::v2::DeltaDiscoveryRequest* getNextRequestInternal();
```

Comment:
This should probably return a copy of the DeltaDiscoveryRequest or a std::unique_ptr.

Reply:
see my comments elsewhere in this PR